### PR TITLE
[WIP] API: clean up attributes and expose .annotate_epochs

### PR DESCRIPTION
**Pull request by dengemann**
Getting started

- [x] propose AP through tests
- [ ] update core code
- [ ] examples

### CODE REVIEW

File: autoreject/autoreject.py [hunk 1]
```diff
@@ -378,7 +378,7 @@ class LocalAutoReject(BaseAutoReject):
     ----------
     epochs : instance of mne.Epochs
         The epochs object
-    consensus_perc : float (0 to 1.0)
+    consensus : float (0 to 1.0)
```
Comment (jasmainak) [inline autoreject/autoreject.py]: float | int ?

File: autoreject/autoreject.py [hunk 2]
```diff
@@ -398,49 +398,37 @@ class LocalAutoReject(BaseAutoReject):
...
-        according to the sensor thresholds.
-    fix_log : array, shape (n_epochs, n_channels)
-        Similar to bad_segments, but with entries 0, 1, and 2.
-            0 : good data segment
-            1 : bad data segment not interpolated
-            2 : bad data segment interpolated
-    bad_epochs_idx : array
-        The indices of bad epochs.
+    picks_ : ndarray, shape(n_channels,)
+        The channels actually used by autoreject.
```
Comment (jasmainak) [inline autoreject/autoreject.py]: The channel indices ...

File: autoreject/autoreject.py [hunk 3]
```diff
@@ -398,49 +398,37 @@ class LocalAutoReject(BaseAutoReject):
...
-                             'You gave me %s.' % consensus_perc)
-        self.consensus_perc = {
-            ch: consensus_perc for ch in ('mag', 'grad', 'eeg')}
-        self.n_interpolate = {
-            ch: n_interpolate for ch in ('mag', 'grad', 'eeg')}
+        if not (0 <= consensus <= 1):
+            raise ValueError('"consensus" must be between 0 and 1. '
+                             'You gave me %s.' % consensus)
+        self.consensus = consensus
+        self.consensus_ = dict()
```
Comment (jasmainak) [inline autoreject/autoreject.py]: I think this should be in LocalAutoRejectCV not here

File: autoreject/autoreject.py [hunk 4]
```diff
@@ -398,49 +398,37 @@ class LocalAutoReject(BaseAutoReject):
...
+        self.n_interpolate = n_interpolate
+        self.n_interpolate_ = dict()
```
Comment (jasmainak) [inline autoreject/autoreject.py]: Same comment here

File: autoreject/autoreject.py [hunk 5]
```diff
@@ -459,22 +447,21 @@ def _vote_bad_epochs(self, epochs):
...
+                                  ch_type, picks, verbose='progressbar'):
         """Interpolate the bad epochs."""
         # 1: bad segment, # 2: interpolated
         fix_log = drop_log.copy()
         ch_names = epochs.ch_names
-        non_picks = np.setdiff1d(range(epochs.info['nchan']), self.picks)
+        non_picks = np.setdiff1d(range(epochs.info['nchan']), picks)
         interp_channels = list()
-        n_interpolate = self.n_interpolate[ch_type]
+        n_interpolate = self.n_interpolate_[ch_type]
```
Comment (jasmainak) [inline autoreject/autoreject.py]: Umm ... not sure I understand this change

File: autoreject/autoreject.py [hunk 6]
```diff
@@ -516,18 +503,20 @@ def _get_bad_epochs(self, bad_sensor_counts, ch_type):
 
         return bad_epochs_idx, sorted_epoch_idx, n_epochs_drop
 
-    def _annotate_epochs(self, threshes, epochs):
+    def _annotate_epochs(self, threshes, epochs, picks):
         """Get essential annotations for epochs given thresholds."""
-        ch_type = _get_ch_type_from_picks(self.picks, epochs.info)[0]
+        ch_type = _get_ch_type_from_picks(
+            picks=self.picks_, info=epochs.info)[0]
```
Comment (jasmainak) [inline autoreject/autoreject.py]: Why self.picks_ and not self.picks? It's not estimated

File: autoreject/autoreject.py [hunk 7]
```diff
@@ -536,6 +525,69 @@ def _annotate_epochs(self, threshes, epochs):
...
+            The epoched data to be annotated.
+        picks : np.ndarray, shape(n_channels, ) | list | None
+            The channel indices to be used. If None, the .picks attribute
+            will be used.
+
+        Returns
+        -------
+        annot : dict
+            fix_log : array, shape (n_epochs, n_channels)
+                Similar to bad_segments, but with entries 0, 1, and 2.
```
Comment (jasmainak) [inline autoreject/autoreject.py]: There is no more bad_segments right?

File: autoreject/autoreject.py [hunk 8]
```diff
@@ -536,6 +525,69 @@ def _annotate_epochs(self, threshes, epochs):
...
+                    0 : good data segment
+                    1 : bad data segment not interpolated
+                    2 : bad data segment interpolated
+            bad_epochs_idx : array
+                The indices of bad epochs.
+            good_epochs_idx : array
+                The indices of good epochs.
+            interp_channels : dict of list of list of str
```
Comment (jasmainak) [inline autoreject/autoreject.py]: A bit too dense. An example may help

File: autoreject/autoreject.py [hunk 9]
```diff
@@ -536,6 +525,69 @@ def _annotate_epochs(self, threshes, epochs):
...
+                The names of the channels to be interpolated
+                for each trial by channel type.
+            picks_by_type : list of tuple
```
Comment (jasmainak) [inline autoreject/autoreject.py]: Not sure I follow ...

File: autoreject/autoreject.py [hunk 10]
```diff
@@ -536,6 +525,69 @@ def _annotate_epochs(self, threshes, epochs):
...
+        else:
+            ch_type = _get_ch_type_from_picks(picks=picks, info=epochs.info)
+            (_, _, interp_channels, fix_log, bad_epochs_idx,
+             good_epochs_idx) = self._annotate_epochs(
+                self, self.threshes_, epochs, picks=picks)
+            annot = dict(
+                fix_log=fix_log,
+                interp_channels={ch_type: interp_channels},
+                bad_epochs_idx=bad_epochs_idx,
+                good_epochs_idx=good_epochs_idx,
```
Comment (jasmainak) [inline autoreject/autoreject.py]: Do we need to return both good_epochs_idx and bad_epochs_idx?

File: autoreject/autoreject.py [hunk 11]
```diff
@@ -536,6 +525,69 @@ def _annotate_epochs(self, threshes, epochs):
...
+
+        .. note::
+           If multiple channel types are present, annot['bad_epochs_idx']
+           reflects the union of bad trials across channel types.
+
+        Parameters
+        ----------
+        epochs : instance of mne.Epochs
+            The epoched data to be annotated.
+        picks : np.ndarray, shape(n_channels, ) | list | None
```
Comment (jasmainak) [inline autoreject/autoreject.py]: shape (n_channels)

File: autoreject/autoreject.py [hunk 12]
```diff
@@ -536,6 +525,69 @@ def _annotate_epochs(self, threshes, epochs):
...
+            The channel indices to be used. If None, the .picks attribute
+            will be used.
+
+        Returns
+        -------
+        annot : dict
+            fix_log : array, shape (n_epochs, n_channels)
```
Comment (jasmainak) [inline autoreject/autoreject.py]: If picks has shape n_channels then this should have shape > n_channels

File: autoreject/autoreject.py [hunk 13]
```diff
@@ -594,46 +641,23 @@ def transform(self, epochs):
...
             self._interpolate_bad_epochs(
-                epochs_out, interp_channels=interp_channels,
-                verbose=self.verbose)
-        if np.any(bad_epochs_idx):
-            epochs_out.drop(bad_epochs_idx, reason='AUTOREJECT')
+                epochs_out, interp_channels=annot['interp_channels'][ch_type],
+                picks=this_picks, verbose=self.verbose)
+
+        if np.any(annot['bad_epochs_idx']):
+            epochs_out.drop(annot['bad_epochs_idx'], reason='AUTOREJECT')
```
Comment (jasmainak) [inline autoreject/autoreject.py]: I'm wondering if this should be inside annotate_epochs and we do away with good_epochs_idx and bad_epochs_idx, at least in the public API?

**Comment by jasmainak** [comment]
@dengemann you need to rebase

**Comment by dengemann** [comment]
Opening new PR ...

**Pull request closed without merging**